libraries/lib-music-information-retrieval/tests/MirTestUtils.h

```diff
@@ -0,0 +1,162 @@
+ */
+template <typename Result>
+RocInfo
+GetRocInfo(std::vector<Result> results, double allowedFalsePositiveRate = 0.)
+{
+   const auto truth = std::mem_fn(&Result::truth);
+   const auto falsity = std::not_fn(truth);
+
+   // There is at least one positive and one negative sample.
+   assert(any_of(results.begin(), results.end(), truth));
+   assert(any_of(results.begin(), results.end(), falsity));
+
+   assert(allowedFalsePositiveRate >= 0. && allowedFalsePositiveRate <= 1.);
+   allowedFalsePositiveRate = std::clamp(allowedFalsePositiveRate, 0., 1.);
+
+   // Sort the results by score, descending.
+   std::sort(results.begin(), results.end(), [](const auto& a, const auto& b) {
+      return a.score > b.score;
+   });
+
+   const auto size = results.size();
+   const auto numPositives = count_if(results.begin(), results.end(), truth);
+   const auto numNegatives = size - numPositives;
+
+   // Find true and false positive rates for various score thresholds.
+   // True positive and false positive counts are nondecreasing with i,
+   // therefore if false positive rate has increased at some i, true positive
+   // rate has not decreased.
+   std::vector<double> truePositiveRates;
+   truePositiveRates.reserve(size);
+   std::vector<double> falsePositiveRates;
+   falsePositiveRates.reserve(size);
+   size_t numTruePositives = 0;
+   size_t numFalsePositives = 0;
+   for (const auto &result : results)
+   {
+      if (result.truth)
+         ++numTruePositives;
+      else
+         ++numFalsePositives;
+      truePositiveRates.push_back(
+         static_cast<double>(numTruePositives) / numPositives);
+      falsePositiveRates.push_back(
+         static_cast<double>(numFalsePositives) / numNegatives);
+   }
+
+   // Now find the area under the non-decreasing curve with FPR as x-axis,
+   // TPR as y, and i as a parameter.  (This curve is within a square with unit
+   // side.)
+   double auc = 0.;
+   for (size_t i = 0; i <= size; ++i)
+   {
+      const auto leftFpr = i == 0 ? 0. : falsePositiveRates[i - 1];
+      const auto rightFpr = i == size ? 1. : falsePositiveRates[i];
+      const auto leftTpr = i == 0 ? 0. : truePositiveRates[i - 1];
+      const auto rightTpr = i == size ? 1. : truePositiveRates[i];
+      const auto trapezoid = (rightTpr + leftTpr) * (rightFpr - leftFpr) / 2.;
+      assert(trapezoid >= 0); // See comments above
+      auc += trapezoid;
+   }
+
+   // Find the parameter at which the x coordinate exceeds the allowed FPR.
+   const auto it = std::upper_bound(
+      falsePositiveRates.begin(), falsePositiveRates.end(),
+      allowedFalsePositiveRate);
+
+   if (it == falsePositiveRates.end())
+      // All breakpoints satify the constraint. Return the least score.
+      return { auc, results.back().score };
+   else if (it == falsePositiveRates.begin())
+      // No breakpoint satisfies the constraint. Return the greatest score.
+      return { auc, results.front().score };
```

Comment:
From @Paul-Licameli:
> I think this branch should return greatest score plus some epsilon.

Reply:
It could, but since the false-positive tolerance itself being rather arbitrary, we shouldn't be preoccupied with that kind of precision.